notifier/notifier.go

```diff
@@ -610,16 +611,79 @@ type alertmanager interface {
 	url() *url.URL
 }
 
-type alertmanagerLabels struct{ labels.Labels }
+type alertmanagerLabels struct {
+	labels           labels.Labels
+	discoveredLabels labels.Labels
+}
 
 const pathLabel = "__alerts_path__"
 
 func (a alertmanagerLabels) url() *url.URL {
 	return &url.URL{
-		Scheme: a.Get(model.SchemeLabel),
-		Host:   a.Get(model.AddressLabel),
-		Path:   a.Get(pathLabel),
+		Scheme: a.labels.Get(model.SchemeLabel),
+		Host:   a.labels.Get(model.AddressLabel),
+		Path:   a.labels.Get(pathLabel),
+	}
+}
+
+type Target struct {
+	discoveredLabels labels.Labels
+	labels           labels.Labels
+}
+
+func NewTarget(labels, discoveredLabels labels.Labels) Target {
+	return Target{labels: labels, discoveredLabels: discoveredLabels}
+}
+
+func (a Target) Labels() labels.Labels {
+	lb := labels.NewBuilder(labels.EmptyLabels())
+	a.labels.Range(func(l labels.Label) {
+		if !strings.HasPrefix(l.Name, model.ReservedLabelPrefix) {
+			lb.Set(l.Name, l.Value)
+		}
+	})
+	url := url.URL{
+		Scheme: a.labels.Get(model.SchemeLabel),
+		Host:   a.labels.Get(model.AddressLabel),
+		Path:   a.labels.Get(pathLabel),
+	}
+	instance := url.String()
+	if instance != "" {
+		lb.Set(model.InstanceLabel, instance)
+	}
+	return lb.Labels()
+}
+
+func (a Target) DiscoveredLabels() labels.Labels {
+	lb := labels.NewBuilder(labels.EmptyLabels())
+	a.discoveredLabels.Range(func(l labels.Label) {
+		lb.Set(l.Name, l.Value)
+	})
+	return lb.Labels()
```

Comment:
I don't follow what this is supposed to achieve.  Why not return the member, or perhaps `.Copy()` ?